### ISSUE TITLE
accessing nested widgets with gaps

### ISSUE THREAD

**Issue by yomed** [raptorjs/marko-widgets#48]
There seem to be some issues in accessing nested widgets when there are gaps between widgets. In this case we have some parentWidget. And then we create children widgets (similar issues can be replicated when using custom tags rather than invoked renderer).

Say that we have some child templates (foo, bar, and baz). Only baz is an actual marko widget, while foo and bar are just template.marko files.
```
data.renderers = [
    require('./foo'),
    require('./bar'),
    require('./baz')
];

```

Then we assign nested widgets as follows (although these are not all "true" widgets):
```
<for each="renderer in data.renderers">
    <invoke function="renderer(myInputObject, out)" w-id="childWidget[]" />
</for>
```

This will result in markup as follows:
```
<div class="foo">foo</div>
<div class="bar">bar</div>
<div class="baz" id="parentWidget[0]-childWidget[2]">baz</div>
```

The childWidget in this case starts at index 2, because foo and bar were correctly skipped. However, when trying to access the nested widgets, this gap seems to cause problems.

```
parentWidget.getWidgets('childWidget') // comes up empty
```

Should everything be created as widget to begin with, in order to avoid issues like this? Wouldn't that create unnecessary overhead? Ideally we would only use w-id when we know it's a widget we want to access, but I imagine in some cases that may not be possible when using both widgets and non-widgets similarly.

**Comment by patrick-steele-idem** [comment]
I think this is fair to say that what you are doing is very much an edge case. When a `w-id` attribute is assigned, we surround the nested widget with code that passes along a generated widget ID to the nested widget by storing the assigned widget ID in the `out` context. When the nested widget renders it will pick up its assigned widget ID out of the same `out` context variable. After the widget renders we clear out that variable in the `out` context. It is actually considered user error if a widget ID is assigned but no widget renders to pick up that ID. We can try to handle your use case, but it will put some more complexity into the runtime... I'll take a closer look and see what we can do.

**Comment by yomed** [comment]
I agree it is somewhat of an edge case. I guess it seems like either the index shouldn't get incremented with non-widgets, or the parent should be able to access child widgets regardless of index gaps. I have had the gap issue come up in the past, and it can usually be resolved by either 1) making everything widgets, or 2) being very strict about when you try to assign w-id(s).

**Comment by patrick-steele-idem** [comment]
I'm looking into a solution that would have the nested widget increment the repeated widget ID counter instead of the parent providing the pre-incremented widget ID counter. I'll provide an update soon.

**Comment by patrick-steele-idem** [comment]
Changing the logic such that the nested widget increments the repeated ID seems to solve your edge case without introducing any other problems. I will be publishing a new version soon.

**Comment by patrick-steele-idem** [comment]
New version published: `marko-widgets@4.2.1`

**Comment by yomed** [comment]
Awesome, thanks!
